### ISSUE TITLE
Begin converting short codes to longs codes

### ISSUE THREAD

**Issue by pedrogit** [edwardsmarc/CASFRI#211]
Many attributes can already adopt a long coding scheme:

STAND_STRUCTURE, DIST_TYPE_X, SOIL_MOIST_REG, PRODUCTIVE_FOR, SITE_CLASS, NAT_NON_VEG, NON_FOR_ANTH, NON_FOR_VEG, WETLAND_TYPE, WET_VEG_COVER, WET_LANDFORM_MOD, ECO_SITE

SPECIES_X should be all uppercase and have an underscode between the genus and the specie. e.g. "POPU_BALB" instead of "Popu balb" In order to consistent with other codes which are all uppercases and a single word without spaces.

We should list long codes proposition here before implementing them...

**Comment by pedrogit** [comment]
All long codes for those attributes have been determined in the following "LongCodes" sheet:

https://docs.google.com/spreadsheets/d/1Hn8jS-VIph1WSk58CmWyb_FAUeOxKFXjUJIg38neLxU/edit?ts=5d94dc9d#gid=1443674988

**Comment by edwardsmarc** [comment]
Converting the species codes could potentially be combined with #191 

Tasks for this issue:
* Update specs with long codes
* Update existing translation tables to use long codes
* Update lookup tables to use long codes
* Use long codes for all new translation tables

**Comment by kdarndt** [comment]
Note that for these codes and for validation (#191), a lookup table associated by foreign key to the appropriate attribute in a CASFRI table provides validation for "free" (built-in to the database).

For example:
```
stand_structure_codes
code  |   description_en | description_fr
------------------------------------------------------
S     | Single layered Stand | ....
M     | Mixed stand          | ....
...

CAS
cas_id | stand_structure | .....
------------------------------------------------------
..     |   S             | .....
..     |   J             | ....         <--- Would not be allowed to be entered
```

The query to add a FK constraint to CAS would look something like this (I haven't tested or checked this, so the syntax might not be quite right):

```
ALTER TABLE cas ADD CONSTRAINT cas_stand_structure_fk (stand_structure) REFERENCES stand_structure_codes (code);
```

This will work for both short and long codes.  The foreign keys could be added after a load to improve the load performance and as part of a "validation" step.

Also, lookup tables could be located in a separate schema, if desired, to keep the CASFRI schema less cluttered.

**Comment by pedrogit** [comment]
What about only defining a constraint enumerating the list of acceptable value instead?

Only the translation script has to be modified, no need to create file based lookup tables and to load them.

**Comment by pedrogit** [comment]
Ok I got a good reason to implement lookup tables: you might want to reuse the table for other purposes (e.g. applying the same constraint on another table derived from CASFRI)

**Comment by pedrogit** [comment]
We could script the creation of all the lookup tables rather then creating file based table and load them. What do you think?

**Comment by kdarndt** [comment]
Yeah, I like lookup tables better because it is easier to modify them (particularly if the same set of values is used to validate multiple attributes) as you already mentioned.

I think a simple .sql script to create all the loop tables is sufficient.  That's typically what I've done with other projects.  It isn't much harder to add a value to an INSERT or COPY statement than to a .csv file.

**Comment by edwardsmarc** [comment]
Is switching the schema to use long codes a high priority at this stage? It would be significant work to make the changes outlined in my comment above. Is this more important than proceeding with the translations?

@prvernier and I just discussed this and made the following observations:
* Translation tables would be much wider. True, the codes will be more readable but the tables themselves will become more cumbersome because the number of characters per cell will grow, a lot in some cases that use long matchList functions.
* If we go ahead, we should meet to complete the table attached above before proceeding. For example, there is currently a disturbance code included that is not in the specs (CL).
* We agree this issue is useful, but we should check this is a priority for all parties before proceeding.

**Comment by pedrogit** [comment]
I don't think the fact that translation tables will become long is a strong argument. They are already long and if they get too long they can be converted to lookup tables. The quality of CASFRI should not be subordinated to our difficulty to read translation tables. In the worst case we could limit the long codes to a certain fixed number of charaters.

If "CL" is not in the specs then it's a bug because it is in the database:

SELECT DISTINCT dist_type_1
FROM casfri50.dst_all;

I don't think we need to meet but some of us can review the proposed long codes and comment. Many are obvious. Some could be changed.

It's such a huge task if you search and replace each code in many files using the right tool and at every change use the tests to make sure everything went fine. I would say one day of work.

**Comment by kdarndt** [comment]
In my opinion, this is a nice-to-have, but not a top priority.  Particularly if there are lookup tables, the description of the code will be easy to find within the database (and of course, the data dictionary documentation will list them all).

**Comment by pedrogit** [comment]
The idea was to make this major change before we implement lookup tables and before you start developing on top of CASFRI so that at least those two developement would not have to be updated after the move to long codes.

**Comment by pedrogit** [comment]
These is also the fact that right now all error codes are long codes and all other codes are short codes. This is odd.

**Comment by kdarndt** [comment]
I think this change will impact writing (and thus re-writing) the translation files more than it would impact anything I do in the short term.  The structure of the schema itself (CAS, LYR, NFL, DST, etc ) won't change (other than adding some lookup tables).

So as far as we're concerned it is not a priority.  But if you're past the point of no return on this, and you can accomplish all the other priorities.....

**Comment by edwardsmarc** [comment]
Note that the code should be OPEN_MUSKEG not OPEN_MUSKED.

**Comment by pedrogit** [comment]
Reoppening as some long codes were missing for BC.

Already fixed but need to check constraint 5.23 on next complete translation.

**Comment by edwardsmarc** [comment]
GRAMMINOIDS should be changed to GRAMINOIDS
